Improve performance of parquet dictionary to domain conversion

raunaqmorarka opened the pull request:
## Description

Improved logic for constructing domain from dictionary values in TupleDomainParquetPredicate by using the faster code path for constructing ValueSet from singleton values rather than Ranges.

Benchmark                                                  Mode  Cnt  Score Before      Score After     Units
BenchmarkTupleDomainParquetPredicate.domainFromDictionary  avgt   10  354.194 ± 29.631  211.831 ± 8.793 ms/op

<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Hive, Hudi, Iceberg, Delta
* Improve performance of reading parquet filter for queries with filters. ({issue}`issuenumber`)
```

findepi commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -695,4 +781,42 @@ private Function<Integer, Object> getConverter(PrimitiveType primitiveType)
         }
     }
+
+    private sealed interface ValuesList
+            permits StatisticsRangeList, DiscreteValuesList
```
`sealed` and `permits` are cool, but pretty redundant, since all these classes are private to this file anyway

findepi commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -695,4 +781,42 @@ private Function<Integer, Object> getConverter(PrimitiveType primitiveType)
+    {
+        private StatisticsRangeList(List<StatisticsRange> rangesList)
+        {
+            this.rangesList = requireNonNull(rangesList, "rangesList is null");
```
why no defensive copy here?

findepi commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -695,4 +781,42 @@ private Function<Integer, Object> getConverter(PrimitiveType primitiveType)
+    {
+        private DiscreteValuesList(Collection<?> values)
+        {
+            this.values = requireNonNull(values, "values is null");
```
why no defensive copy here?

findepi commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -244,16 +243,17 @@ public static Domain getDomain(
-            boolean hasTrueValues = minimums.stream().anyMatch(value -> (boolean) value) || maximums.stream().anyMatch(value -> (boolean) value);
-            boolean hasFalseValues = minimums.stream().anyMatch(value -> !(boolean) value) || maximums.stream().anyMatch(value -> !(boolean) value);
+            if (valuesList instanceof DiscreteValuesList) {
+                throw new VerifyException("Encountered dictionary encoding for boolean");
```
why not?

raunaqmorarka replied:
Parquet encoding does not support dictionary for booleans (see `ParquetEncoding#initDictionary`)

findepi commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -268,137 +268,224 @@ private static Domain getDomain(
+                        Object max = statisticsRange.max();
 
-                    long minValue = min instanceof Binary ? getShortDecimalValue(((Binary) min).getBytes()) : asLong(min);
-                    long maxValue = min instanceof Binary ? getShortDecimalValue(((Binary) max).getBytes()) : asLong(max);
```
there was a bug here, right?

can you please fix it as a prep commit?

raunaqmorarka replied:
It's not possible for the min and max values to be of different types in the file, it's more of a consistency/readability fix

findepi commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -268,137 +268,224 @@ private static Domain getDomain(
+            for (Object value : discreteValuesList.values()) {
+                values.add(Int128.fromBigEndian(((Binary) value).getBytes()));
+            }
+            return Domain.create(ValueSet.copyOf(type, values), hasNullValue);
```
I think you could have a bit less code if you checked `valuesList` first and then `decimal.isShort()`
the
```
List<Object> values = new ArrayList<>(valuesList.size());
...
return Domain.create(ValueSet.copyOf(type, values), hasNullValue);
```

would be shared.

raunaqmorarka replied:
Trying to share that makes it slightly harder to read because I have to
```
if (decimal.isShort()) {
  .....
}
else {
  .....
}
```
instead of
```
if (decimal.isShort()) {
  ......
}
....
```

findepi commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -695,4 +781,42 @@ private Function<Integer, Object> getConverter(PrimitiveType primitiveType)
+        }
+    }
+
+    record StatisticsRange(Object min, Object max) {}
```
are `min` and `max` nullable?

findepi reviewed (commented):
LGTM

Quite considerable code duplication. is it worth it?

sopel39 commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -268,137 +268,224 @@ private static Domain getDomain(
+                    }
+                    ranges.add(Range.range(type, min, true, max, true));
                 }
-
```
undo

sopel39 commented on lib/trino-parquet/src/main/java/io/trino/parquet/predicate/TupleDomainParquetPredicate.java:
```diff
@@ -268,137 +268,224 @@ private static Domain getDomain(
-                ranges.add(Range.range(type, min, true, max, true));
+                return Domain.create(ValueSet.ofRanges(ranges), hasNullValue);
             }
 
```
undo

sopel39 reviewed (approved):
lgtm % @findepi comments %

`SortedRangeSet#of(io.trino.spi.type.Type, java.util.Collection<?>)` and `SortedRangeSet#buildFromUnsortedRanges` look similar (comparisons etc). Maybe it's possible to improve range based method

raunaqmorarka commented:
> lgtm % @findepi comments %
> 
> `SortedRangeSet#of(io.trino.spi.type.Type, java.util.Collection<?>)` and `SortedRangeSet#buildFromUnsortedRanges` look similar (comparisons etc). Maybe it's possible to improve range based method

I looked into what is slow here
Before 
<img width="1706" alt="Screenshot 2022-12-02 at 12 51 23 PM" src="https://user-images.githubusercontent.com/4344846/205240087-ae9246ff-c939-4793-8969-cc782db9c621.png">

After
<img width="1577" alt="Screenshot 2022-12-02 at 12 59 49 PM" src="https://user-images.githubusercontent.com/4344846/205240111-17d65037-e77b-4440-9fda-a13beb3e32ec.png">

The problem is that `Range#range` performs a lookup of comparison operator for the type. This gets repeated unnecessarily for every range object.
Same problem was addressed in https://github.com/trinodb/trino/pull/10239 as well.
I've changed the approach here to use a SortedRanges builder which caches comparison operator and avoids that bottleneck. While this is still not exactly as fast the previous approach it's still significantly faster and far less code so I've simplified the PR now.

sopel39 commented on core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java:
```diff
@@ -981,6 +981,44 @@ private String formatRanges(ConnectorSession session, int limit)
+
+        private Builder(Type type, int expectedSize)
+        {
+            this.type = type;
```
rnn

sopel39 commented on core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java:
```diff
@@ -981,6 +981,44 @@ private String formatRanges(ConnectorSession session, int limit)
+    {
+        private final Type type;
+        private final MethodHandle rangeComparisonOperator;
+        private final List<Range> ranges;
```
nit: We could probably optimize it further by using `BlockBuilder` directly
